nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java

```diff
@@ -0,0 +1,266 @@
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package com.sun.enterprise.v3.admin;
+
+import com.sun.enterprise.util.OS;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.lang.ProcessBuilder.Redirect;
+import java.lang.System.Logger;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.logging.Handler;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.glassfish.main.jul.GlassFishLogManager;
+
+import static java.lang.System.Logger.Level.ERROR;
+import static java.lang.System.Logger.Level.INFO;
+
+
+/**
+ * Starts the server after it is completely turned off.
+ * <p>
+ * Shutdown hooks are asynchronously executed - we need to wait for the end
+ * all GlassFish shutdown hooks, at least of those we know.
+ * <p>
+ * All shutdown hooks are executed in parallel including this thread.
+ * As we don't have any direct reference to other hooks, but we need to
+ * shut GlassFish down before we can start it, we postpone our part
+ * of the job as much as possible.
+ */
+class StartServerShutdownHook extends Thread {
+
+    private static final Logger LOG = System.getLogger(StartServerShutdownHook.class.getName());
+    private static final boolean LOG_RESTART = Boolean.parseBoolean(System.getenv("AS_RESTART_LOGFILES"));
+    private static final Path LOGDIR = new File(System.getProperty("com.sun.aas.instanceRoot"), "logs").toPath()
+        .toAbsolutePath();
+    private static final Predicate<Thread> FILTER_OTHER_HOOKS = t -> t.getName().startsWith("GlassFish")
+        && t.getName().endsWith("Shutdown Hook");
+
+    private final PrintStream logFile;
+    private final ProcessBuilder builder;
+    private final Instant startTime;
+
+    StartServerShutdownHook(String classpath, String[] sysProps, String classname, String[] args) {
+        super("GlassFish Restart Shutdown Hook");
+        setDaemon(false);
+        if (classname == null || classname.isBlank()) {
+            throw new IllegalArgumentException("classname was null");
+        }
+        this.startTime = Instant.now();
+        this.logFile = getLogFileOld(startTime);
+        this.builder = prepareStartup(startTime, classpath, sysProps, classname, args);
+    }
+
+
+    @Override
+    public void run() {
+        try {
+            // We do want to be the last thread alive.
+            Thread.onSpinWait();
```

Comment:
Is spin waiting neccessary here?

Reply:
Yes, originally I implemented your suggestion with onExit, but it freezes on Jenkins - it heavily depends on the OS reaper but not all systems have them.
I added some documentation for that and link to jdk jira. I had yet few others to Stackoverflow, and spent two says with that.
Locally it worked well, but failed on jenkins. 

Maybe you will find something I missed, I don't know.

Reply:
Thank you for answers. I have no more questions. Great job!